wordpress-mobile/WordPress-iOS
# Improves Jetpack Connect through Magic Links.

WPKit PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/320
WPAuth PR: https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/pull/550

Our current logic for handling magic links in WPiOS if flawed.  We store the login / signup state in `NSUserDefaults` and retrieve it once the user taps on the magic link and re-launches the App.

This works fine if the user taps on the correct magic link, but if the user taps on a different magic link, this mechanism is no longer good.

This PR modifies the Jetpack connect flow, so that we no longer store any information in `NSUserDefaults` and instead rely on information present in each magic link to know if we're connecting through Jetpack.

Ref: p4a5px-2DG-p2

## Known Issues

Jetpack connect is very broken.  If you log out of the newly connected account, you'll notice the App kind of breaks.

I'm checking into this - but this issue is also present in `develop` and was not introduced by this PR.

## Testing

Place a breakpoint [here](https://github.com/wordpress-mobile/WordPressAuthenticator-iOS/blob/993be9fccd7a1edf7749564ebefb4fdb0f3e158a/WordPressAuthenticator/Authenticator/WordPressAuthenticator.swift#L335), and check that in each test case the `jetpackLogin` value is set correctly.

### Test 1: Regular Login

1. Be logged out
2. Enter an email for an existing account that can login using magic links
3. Tap “Get a login link by email”
4. Go to your email and tap on the magic link button.
5. Once the breakpoint is hit make sure `jetpackLogin == false`

### Test 2: Regular Signup

1. Be logged out
2. Enter an email for a new account
3. When the “Sign Up” screen comes up tap “Send Link by Email”
4. Go to your email and tap on the magic link button.
5. Once the breakpoint is hit make sure `jetpackLogin == false`

### Test 3: Jetpack Connect

The recommended setup for this test is:

a. Make sure you have two WPCom accounts (A and B).
b. Create a self-hosted site with jurassic.ninja
c. Connect the self-hosted site to WPcom account B through Jetpack (use the web to connect it)

Steps:

1. Be logged out
2. Log in completely with existing account A.
3. Add the self-hosted site (don’t use the WPCom credentials… use the self-hosted ones).
4. The Site Details screen for the self-hosted site will be shown.  Tap on “Stats”.
5. You should be prompted to “Log In”.
6. The email shown should be the right one, click “Continue”.
7. Tap on “Get a login link by email”.
8. Go to your email and tap on the magic link button.
9. Once the breakpoint is hit make sure `jetpackLogin == true`

### Test 4: Double login attempt cancelled.

The recommended setup for this test is:

a. Make sure you have two WPCom accounts (A and B).

Steps:

1. Be logged out
2. Request a magic link for account A but don’t login.
3. Login to account B
4. Go back to your email and tap on the magic link for account A.

You should see that the app is opened but the login attempt ignored.

## PR submission checklist:
